AWS Spring cloud map support (with AWS SDK v2)

hariohmprasath opened the pull request:
Support for integrating with AWS Cloudmap

## :loudspeaker: Type of change
<!--- Put an `x` in the boxes that apply -->
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring


## :scroll: Description
<!--- Describe your changes in detail -->
This pull request adds support for integrating spring boot application with AWS cloudmap. Here is an example of how an simple properties file integration looks like for both cloudmap registration and discovery

```
spring.cloud.aws.cloudmap.region=us-east-1
spring.cloud.aws.cloudmap.enabled=true
spring.application.name=cloudmap-namespace-here

# Discover existing cloudmap instances
spring.cloud.aws.cloudmap.discovery.failFast=false
spring.cloud.aws.cloudmap.discovery.discoveryList[0].service=TestService
spring.cloud.aws.cloudmap.discovery.discoveryList[0].nameSpace=ECS-CloudMap

# Register new instance
spring.cloud.aws.cloudmap.registry.description=Namespace for sample cloudmap registry service
spring.cloud.aws.cloudmap.registry.port=80
spring.cloud.aws.cloudmap.registry.service=a-service
spring.cloud.aws.cloudmap.registry.nameSpace=a-namespace
```

The property file is optional, any application can easily register themselves to cloudmap by adding `spring-cloud-starter-aws-cloudmap` dependency to `pom` or `gradle` files, here is a example:

```xml
<dependency>
  <groupId>io.awspring.cloud</groupId>
  <artifactId>spring-cloud-starter-aws-cloudmap</artifactId>
</dependency>
```


## :bulb: Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Adds support for cloudmap integration with spring boot applications for both service registry and discovery #5 

## :green_heart: How did you test it?
I executed the following tests:

* Ran unit test cases and made sure nothing failed
* Deployed the sample spring boot app checked in part of this pull request in ECS, EKS fargate instances. The application could successfully discover and register the application to AWS cloudmap.

## :pencil: Checklist
<!--- Put an `x` in the boxes that apply -->
- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] I updated reference documentation to reflect the change
- [x] All tests passing
- [x] No breaking changes


## :crystal_ball: Next steps
Code review and merging code to 3.x branch

maciejwalkowiak commented:
Thanks @hariohmprasath. I'll do some polishing and very likely come back to you with some questions. In meantime, there are 3 things missing:

- reference documentation
- infrastructure for running a sample - or an instruction how to run the sample (I guess it's not possible to run it with Localstack)
- nullability annotations & `package-info.java`.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/CloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,81 @@
+	@Bean
+	@ConditionalOnMissingBean
+	CloudMapAutoRegistration createAutoRegistration(ServiceDiscoveryClient serviceDiscovery) {
+		return new CloudMapAutoRegistration(context, serviceDiscovery, properties.getRegistry());
```
Instead of `ApplicationContext`, `CloudMapRegistration` should use `ApplicationEventPublisher` for publishing events. 

Since it's the only bean that uses is, `ApplicationEventPublisher` should be provided as a bean factory method parameter instead of a field in the class.

hariohmprasath replied:
I might need some help on this, do you have a sample for this?

maciejwalkowiak replied:
I pushed a fix.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/CloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,81 @@
+@EnableConfigurationProperties(CloudMapProperties.class)
+@ConditionalOnClass({ ServiceDiscoveryClient.class, ServiceRegistration.class, CloudMapAutoRegistration.class })
+@ConditionalOnProperty(prefix = CloudMapProperties.CONFIG_PREFIX, name = "enabled", matchIfMissing = true)
+public class CloudMapBootstrapConfiguration {
```
Better name would be `CloudMapAutoConfiguration`

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/CloudMapBootstrapConfiguration.java:
```diff
@@ -0,0 +1,81 @@
+ */
+@Configuration(proxyBeanMethods = false)
+@EnableConfigurationProperties(CloudMapProperties.class)
+@ConditionalOnClass({ ServiceDiscoveryClient.class, ServiceRegistration.class, CloudMapAutoRegistration.class })
```
Since `CloudMapAutoRegistration` is in the same jar as the autoconfiguration class, it's always available on the classpath - no need to include it in `@ConditionalOnClass`.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/CloudMapProperties.java:
```diff
@@ -0,0 +1,120 @@
+
+	private CloudMapDiscovery discovery;
+
+	private String region;
```
Region is already included in `AwsClientProperties`.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/CloudMapProperties.java:
```diff
@@ -0,0 +1,120 @@
+	/**
+	 * Overrides the default endpoint.
+	 */
+	private URI endpoint;
```
Endpoint is already included in `AwsClientProperties`.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/CloudMapProperties.java:
```diff
@@ -0,0 +1,120 @@
+	 */
+	private URI endpoint;
+
+	private boolean enabled;
```
No need for `enabled` field.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/CloudMapProperties.java:
```diff
@@ -0,0 +1,120 @@
+
+	private boolean enabled;
+
+	private String annotationBasePackage;
```
This does not seem to be used..?

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/CloudMapProperties.java:
```diff
@@ -0,0 +1,120 @@
+
+	private String annotationBasePackage;
+
+	private Environment environment;
```
This does not seem to be used..?

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/CloudMapProperties.java:
```diff
@@ -0,0 +1,120 @@
+	 */
+	public static final String CONFIG_PREFIX = "spring.cloud.aws.cloudmap";
+
+	private CloudMapRegistryProperties registry;
```
Nested properties classes should be annotated with `@NestedConfigurationProperty` for correct documentation generation.

maciejwalkowiak commented on spring-cloud-aws-autoconfigure/src/main/java/io/awspring/cloud/autoconfigure/cloudmap/properties/discovery/CloudMapDiscovery.java:
```diff
@@ -0,0 +1,54 @@
+
+	private List<CloudMapDiscoveryProperties> discoveryList;
+
+	public boolean isFailFast() {
```
Does not seem to be used..?

maciejwalkowiak commented on spring-cloud-aws-cloudmap/pom.xml:
```diff
@@ -0,0 +1,72 @@
+			<artifactId>spring-web</artifactId>
+		</dependency>
+
+		<dependency>
```
As far as I can see the dependency for webflux exists only because of the web client? If so, `RestTemplate` should be used instead as it's already included in `spring-web`.

maciejwalkowiak commented on spring-cloud-aws-cloudmap/pom.xml:
```diff
@@ -0,0 +1,72 @@
+	<name>Spring Cloud Cloud Map</name>
+	<description>Spring Cloud AWS Cloud Map</description>
+
+	<dependencies>
```
These dependencies are not used in `spring-cloud-aws-cloudmap` module. Since they are required by the autoconfiguration that is triggered only if these dependencies are on the classpath, they should go to starter module and autoconfiguration should trigger only if each of the required dependency is on the classpath.

hariohmprasath replied:
Sorry can you please elaborate the change, I am not able to understand the recommendation here

maciejwalkowiak replied:
Module should have only dependencies used in this very particular module.

Starter should list all the dependencies that are used additionally in autoconfiguration.

Autoconfiguration should be have `@ConditionalOnClass` listing classes from each dependency that is required for the integration to work.

maciejwalkowiak commented on spring-cloud-aws-samples/spring-cloud-aws-cloud-map-sample/src/main/resources/bootstrap.properties:
```diff
@@ -0,0 +1,15 @@
+spring.cloud.aws.cloudmap.region=us-east-1
```
Shouldn't these properties just go to `application.properties`? Or is there a benefit of keeping them in `bootstrap.properties` that I am not aware of?

maciejwalkowiak reviewed (changes requested):
There are quite a few unused fields that I am not sure if just should be deleted or some parts of implementation is still missing.

Missing integration tests stopped me from doing further refactoring. If we can't test against Localstack, we should mock AWS API and perhaps use Wiremock to stub EC2/ECS endpoints.

I feel like the `CloudMapUtils` class is the center of this implementation and ideally we would refactor it to avoid having so much logic in the util class - but this can or even has to wait until we have integration tests so that we can refactor safely.

@hariohmprasath would you be willing to address these issues? In case I am missing something, your opinion is appreciated.

hariohmprasath commented:
> There are quite a few unused fields that I am not sure if just should be deleted or some parts of implementation is still missing.
> 
> Missing integration tests stopped me from doing further refactoring. If we can't test against Localstack, we should mock AWS API and perhaps use Wiremock to stub EC2/ECS endpoints.
> 
> I feel like the `CloudMapUtils` class is the center of this implementation and ideally we would refactor it to avoid having so much logic in the util class - but this can or even has to wait until we have integration tests so that we can refactor safely.
> 
> @hariohmprasath would you be willing to address these issues? In case I am missing something, your opinion is appreciated.

Thanks for reviewing the code @maciejwalkowiak, let me start addressing some of the comments and we can sync up based on the changes

hariohmprasath commented:
> Reference documentation added, package-info with nullability annotation added, no infrastructure required if cloudmap namespace or service doesnt exist the cloudmap connector module will automatically create them

hariohmprasath commented:
> > There are quite a few unused fields that I am not sure if just should be deleted or some parts of implementation is still missing.
> > Missing integration tests stopped me from doing further refactoring. If we can't test against Localstack, we should mock AWS API and perhaps use Wiremock to stub EC2/ECS endpoints.
> > I feel like the `CloudMapUtils` class is the center of this implementation and ideally we would refactor it to avoid having so much logic in the util class - but this can or even has to wait until we have integration tests so that we can refactor safely.
> > @hariohmprasath would you be willing to address these issues? In case I am missing something, your opinion is appreciated.
> 
> Thanks for reviewing the code @maciejwalkowiak, let me start addressing some of the comments and we can sync up based on the changes

@maciejwalkowiak, The last commit that I pushed in has code changes to mock the required HTTP responses. So we can use these tests as a reference for further refactoring. Since CloudMap is not supported in `CloudMap` you may need to connect with AWS using credentials to run this sample.

It just requires the following environment variables to be set before running the sample application:
```properties
AWS_ACCESS_KEY_ID=<key>
AWS_REGION=us-east1
AWS_SECRET_ACCESS_KEY=<secret>
DEPLOYMENT_PLATFORM=ECS
```

Update `CloudMapUtils.getEcsRegistrationAttributes()`, comment out line between 536 to 547 and return hard-coded `ipAddress` and `vpcId`, this is all you need to pretty much to run this sample.

maciejwalkowiak commented:
@hariohmprasath thanks for an update. We still need to have an integration test that checks everything all together. Something like `ParameterStoreConfigDataLoaderIntegrationTests` - does not need to use `SpringApplication`, can just use Spring testing framework capabilities, but it has to give us confidence that when details change, the integration still works. 

Considering it's not possible to use Localstack for that, either we need to mock responses from AWS or run the test against real AWS and spin up resources with Cloudformation. I can imagine this is quite a bit of work but we just cannot merge the integration lacking solid test coverage.

hariohmprasath commented:
> @hariohmprasath thanks for an update. We still need to have an integration test that checks everything all together. Something like `ParameterStoreConfigDataLoaderIntegrationTests` - does not need to use `SpringApplication`, can just use Spring testing framework capabilities, but it has to give us confidence that when details change, the integration still works.
> 
> Considering it's not possible to use Localstack for that, either we need to mock responses from AWS or run the test against real AWS and spin up resources with Cloudformation. I can imagine this is quite a bit of work but we just cannot merge the integration lacking solid test coverage.

Hi @maciejwalkowiak,
Thanks for reaching out to `localstack` team regarding the support. I had a chance to catch up on their response and based on that I have submitted an enhancement (https://github.com/testcontainers/testcontainers-java/issues/5838) + PR (https://github.com/testcontainers/testcontainers-java/pull/5839) to support `AWS CloudMap`  for `localstack` module in `testcontainers` project. 

Once the enhancement and PR gets approved I can use the same to add the required integration test cases. Hope this works for you, let me know your thoughts. Thanks

maciejwalkowiak commented:
@hariohmprasath sounds like the way to go!

hariohmprasath commented:
Hi @maciejwalkowiak, 
I have added integration tests to test service discovery and registration for both the platforms (`ECS` and `EKS`). 
* `EcsCloudMapIntegrationTest.java` - Integration tests for `ECS` platform
* `EksCloudMapIntegrationTest.java` - Integration tests for `EKS` platform

> **Note:** Since `CloudMap` service is only enabled for pro version of `LocalStack` make sure to create an **API Key** (either using trial mode or buying a subscription), update the key in `IntegrationTestUtil.LOCAL_STACK_API_KEY` variable before running the integration tests.

Hopefully this should make things easier for testing, let me know if you want me to update anything. Thanks and appreciate your support for this PR.

hariohmprasath commented:
Hi @maciejwalkowiak, Just checking to see whether you had a chance to review the changes, let me know if you have any questions/suggestions. Thanks

maciejwalkowiak commented:
@hariohmprasath not yet unfortunately but it's still high in my priority list so don't worry - it won't be ignored :-)

hariohmprasath commented:
Hi @maciejwalkowiak,
Hope you are doing great. Any tentative timeline when this support will be released? Thanks

maciejwalkowiak commented:
@hariohmprasath apologies again, but we didn't find enough time to include it in 3.0. 

3.0 will be released in April, CloudMap is planned for 3.1, so Q2/Q3 2023.

hariohmprasath commented:
Sounds good, thanks for letting me know

hariohmprasath commented:
Hi @maciejwalkowiak, Would love to get this merged by early 2024 :)

codefromthecrypt commented on spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/cloudmap/integration/IntegrationTestUtil.java:
```diff
@@ -0,0 +1,121 @@
+	public static final String REGION = "us-east-1";
+	public static final String CIDR_BLOCK = "10.0.0.0/16";
+	public static final String META_DATA_MOCK_RESPONSE = "https://5qjz8e33hj.api.quickmocker.com";
+	public static final String LOCAL_STACK_API = "LOCALSTACK_API_KEY";
```
since this PR, the key is now LOCALSTACK_AUTH_TOKEN

codefromthecrypt commented on spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/cloudmap/integration/IntegrationTestUtil.java:
```diff
@@ -0,0 +1,121 @@
+	public static final String CIDR_BLOCK = "10.0.0.0/16";
+	public static final String META_DATA_MOCK_RESPONSE = "https://5qjz8e33hj.api.quickmocker.com";
+	public static final String LOCAL_STACK_API = "LOCALSTACK_API_KEY";
+	public static final String LOCAL_STACK_API_KEY = ""; // TODO: Add your localstack api key here
```
probably wanna read from env and skip if not present

codefromthecrypt commented on spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/cloudmap/integration/EcsCloudMapIntegrationTest.java:
```diff
@@ -0,0 +1,177 @@
+	// Local stack container with Cloud Map and Ec2 services enabled
+	@Container
+	private static final LocalStackContainer localStackContainer = new LocalStackContainer(
+		DockerImageName.parse("localstack/localstack:1.1.0"))
```
doesn't this need to be localstack-pro? I tried to use normal and it didn't work until I switched the image name. Maybe that changed since this PR was raised (current version is 3.2.0)

codefromthecrypt reviewed (commented):
I saw this and think it is kindof cool, but it will be hard for any community to support unless localstack has a free OSS account. Do you have a relationship with them to ask? I think even if there was such an account it may imply test after merge to prevent key hijacking on pull requests.

I say this not knowing how other amazon resources are managed, so take this advice knowing it is a random drive by.

codefromthecrypt commented on spring-cloud-aws-autoconfigure/src/test/java/io/awspring/cloud/autoconfigure/cloudmap/integration/EksCloudMapIntegrationTest.java:
```diff
@@ -0,0 +1,177 @@
+	 * Register a service with no specification provided in application.properties.
+	 */
+	@Test
+	void registerEksContainerWithCloudMapWithNoSpecification() {
```
curios on this.. ECS makes immediate sense as at the pure container layer it is basically just DNS. On the other hand EKS is in k8s, so there's services, endpoints etc.

what's your opinion on registration here clashing or otherwise being redundant with k8s (endpoint, service, etc) or layers over it (istio)? Is this to support multi-clusters? or is the idea to duplicate register for those using raw cloud map?

curious your thinking as I'm a cloudmap noob and missing the insight on this.

zampettim commented:
We are interested in using this integration as well. What needs to be done to get this reviewed and merged, and how could we help with that process?

hariohmprasath commented:
Hi @maciejwalkowiak, The PR was created on Sep 4, 2022, probably 2 years back :) Any plans to merge it part of a mainline or release?

Vostan commented:
+1

maciejwalkowiak commented:
Again apologies for postponing it for so long. I'll get some support and we will get it through.
